sgl-kernel/csrc/moe/moe_sum_reduce.cu

```diff
@@ -0,0 +1,303 @@
+
+  TORCH_CHECK(input.is_contiguous(), "expect input to be contiguous");
+  TORCH_CHECK(output.is_contiguous(), "expect output to be contiguous");
+
+  const int64_t token_num = input.size(0);
+  const int64_t topk_num = input.size(1);
+  const int64_t hidden_dim = input.size(2);
+
+  const int64_t in_stride_token = input.stride(0);
+  const int64_t in_stride_topk = input.stride(1);
+  const int64_t out_stride_token = output.stride(0);
+
+  const float scale = static_cast<float>(routed_scaling_factor);
+
+  auto stream = at::cuda::getCurrentCUDAStream();
+
+  const bool fast_bf16_vec_ok = (input.scalar_type() == at::kBFloat16) && (token_num > 256) && (hidden_dim % 8 == 0);
+
+  // Fast path for bf16 vectorize
+  if (fast_bf16_vec_ok) {
+    constexpr int WARPS_PER_BLOCK = 8;
+    constexpr int THREADS = WARPS_PER_BLOCK * 32;
+
+    const int64_t n_chunks = hidden_dim / 8;
+    int64_t grid_x = (n_chunks + 32 - 1) / 32;
+    if (grid_x > 65535) grid_x = 65535;
+
+    int64_t grid_y = (token_num + WARPS_PER_BLOCK - 1) / WARPS_PER_BLOCK;
+    if (grid_y > 65535) grid_y = 65535;
+
+    dim3 block(THREADS);
+    dim3 grid(static_cast<unsigned>(grid_x), static_cast<unsigned>(grid_y));
+
+    auto stream = at::cuda::getCurrentCUDAStream();
+
+    moe_sum_reduce_warp_per_token_vec_kernel<WARPS_PER_BLOCK><<<grid, block, 0, stream>>>(
+        reinterpret_cast<const at::BFloat16*>(input.data_ptr<at::BFloat16>()),
+        reinterpret_cast<at::BFloat16*>(output.data_ptr<at::BFloat16>()),
+        token_num,
+        hidden_dim,
+        topk_num,
+        in_stride_token,
+        in_stride_topk,
+        out_stride_token,
+        scale);
+
+    TORCH_CHECK(cudaGetLastError() == cudaSuccess, "moe_sum_reduce CUDA kernel launch failed");
+    return;
+  }
+
+  const bool use_warp_token = (token_num > 128);
+
+  auto dispatch_topk = [&](auto&& launch_kernel) {
+    switch (topk_num) {
+      case 2:
+        launch_kernel(std::integral_constant<int, 2>{});
+        break;
+      case 4:
+        launch_kernel(std::integral_constant<int, 4>{});
+        break;
+      case 8:
+        launch_kernel(std::integral_constant<int, 8>{});
+        break;
+      case 9:
+        launch_kernel(std::integral_constant<int, 9>{});
+        break;
+      default:
+        launch_kernel(std::integral_constant<int, -1>{});
+        break;
+    }
+  };
+
+  if (!use_warp_token) {
+    // ---------- small-token ----------
+    const int block_size = 256;
+    int64_t grid_x = (hidden_dim + block_size - 1) / block_size;
+    grid_x = grid_x > 65535 ? 65535 : grid_x;
+    int64_t grid_y = token_num < 65535 ? token_num : 65535;
+
+    dim3 block(block_size);
+    dim3 grid(static_cast<unsigned>(grid_x), static_cast<unsigned>(grid_y));
+
+    AT_DISPATCH_FLOATING_TYPES_AND2(
+        at::kHalf, at::kBFloat16, input.scalar_type(), "moe_sum_reduce_cuda_small_token", [&] {
+          using scalar_t_ = scalar_t;
+
+          auto launch_small = [&](auto topk_c) {
```

Comment:
`launch_small` can be renamed to `lauch_small_token_kernel` ?

Reply:
Renamed.